sklearn/calibration.py

```diff
@@ -25,6 +27,230 @@
+    thresholds.
+
+    If cv="prefit" the base estimator is assumed to be fitted and all data will
+    be used for the selection of the cutoff point. Otherwise the decision
+    threshold is calculated as the average of the thresholds resulting from the
+    cross-validation loop.
+
+    Parameters
+    ----------
+    base_estimator : instance BaseEstimator
+        The classifier whose decision threshold will be adapted according to
+        the acquired cutoff point
+
+    method : str
+        The method to use for choosing the cutoff point.
+
+        - 'roc', selects the point on the roc_curve that is closer to the ideal
+        corner (0, 1)
+
+        - 'max_tpr', selects the point that yields the highest true positive
+        rate with true negative rate at least equal to the value of the
+        parameter min_val_tnr
+
+        - 'max_tnr', selects the point that yields the highest true negative
+        rate with true positive rate at least equal to the value of the
+        parameter min_val_tpr
+
+    pos_label : object
+        Object representing the positive label
+        (default value: 1)
+
+    cv : int, cross-validation generator, iterable or "prefit" (optional)
+        Determines the cross-validation splitting strategy. If cv="prefit" the
+        base estimator is assumed to be fitted and all data will be used for
+        the calibration of the probability threshold
+        (default value: "prefit")
+
+    min_val_tnr : float in [0, 1]
+        In case method = 'max_tpr' this value must be set to specify the
+        minimum required value for the true negative rate
+
+    min_val_tpr : float in [0, 1]
+        In case method = 'max_tnr' this value must be set to specify the
+        minimum required value for the true positive rate
+
+    Attributes
+    ----------
+    threshold_ : float
+        Decision threshold for the positive class. Determines the output of
+        predict
+
+    References
+    ----------
+    .. [1] Receiver-operating characteristic (ROC) plots: a fundamental
+           evaluation tool in clinical medicine, MH Zweig, G Campbell -
+           Clinical chemistry, 1993
+
+    """
+    def __init__(self, base_estimator, method='roc', pos_label=1, cv=3,
+                 min_val_tnr=None, min_val_tpr=None):
+        self.base_estimator = base_estimator
+        self.method = method
+        self.pos_label = pos_label
+        self.cv = cv
+        self.min_val_tnr = min_val_tnr
+        self.min_val_tpr = min_val_tpr
+
+    def fit(self, X, y):
+        """Fit model
+
+        Parameters
+        ----------
+        X : array-like, shape (n_samples, n_features)
+            Training data
+
+        y : array-like, shape (n_samples,)
+            Target values. There must be two 2 distinct values.
+
+        Returns
+        -------
+        self : object
+            Instance of self.
+        """
+        if not isinstance(self.base_estimator, BaseEstimator):
```

Comment:
If this is very nearly identical to CalibratedClassifierCV.fit, perhaps we could have a base class or a helper function to avoid repeating the code

Reply:
are you referring parts of the code that deal with handling of the high level logic like validation checks and cv iteration? There is definitely a similar flow there; but I am not sure how beneficial it would be to try to put those in a different method in terms of simplicity. 

Or you thinking of some generic interface; something like a CalibratorMixin?

Reply:
Hey @jnothman I think I addressed all your comment except this one. did my comment above make sense ? Is this what you have in mind?

Reply:
It wouldn't be a mixin as it's unlikely possible to just mix in all sorts of places, but it could be a BaseCalibration. But if you don't think there's enough shared, fair enough.

I don't think we should be using `isinstance(self.base_estimator, BaseEstimator)`, btw. We don't require estimator developers to use BaseEstimator, only to maintain its features. In this case you don't even seem to need full BaseEstimator implementation (i.e. no need for `set_params`), only supporting `fit` and supporting `predict_proba` after fitting (which cannot be tested before fitting).